lib/FrontendTool/FrontendTool.cpp

```diff
@@ -1290,10 +1290,16 @@ static bool performCompile(CompilerInstance &Instance,
   if (opts.PrintClangStats && Context.getClangModuleLoader())
     Context.getClangModuleLoader()->printStatistics();
 
-  emitReferenceDependenciesForAllPrimaryInputsIfNeeded(Instance);
   emitSwiftRangesForAllPrimaryInputsIfNeeded(Instance);
   emitCompiledSourceForAllPrimaryInputsIfNeeded(Instance);
 
+  SWIFT_DEFER {
+    // We might have freed the ASTContext already, but in that case we must have
+    // emitted the dependencies first.
+    if (Instance.hasASTContext())
+      emitReferenceDependenciesForAllPrimaryInputsIfNeeded(Instance);
```

Comment:
I'm surprised this has a dependency on the ASTContext. I tried to remove the explicit ones in a082e9de26407bf428cf61efccc877d098b427d2 since the ASTContext takes a reference to the diagnostic engine on the compiler instance.

Reply:
oh, if there's no dependency on the ASTContext this becomes even simpler.